docs/website/versioned_docs/version-3.0.0/development/devfile.md

```diff
@@ -0,0 +1,709 @@
+
+Below are `odo`-specific examples regarding the Devfile reference. Everything we have listed below is implemented in the current version of `odo`.
+
+This API Reference uses examples and snippets to help you create your own `devfile.yaml` file for your project.
+
+For this reference, we use one singular (working) Node.js example which can be ran from your own environment.
+
+### How `odo` Creates Resources from Devfile
+
+When `odo dev` is ran, `odo` creates the following Kubernetes resources from `container` (component) and `exec` (command) sections of the Devfile:
+* Deployment for running the containers
+* Service for accessability 
+
+### File Synchronization
+
+We use the "inner-loop" process intensively in order to propagate and see changes immediately so you spend less time deploying, and more time writing code. 
+
+Below is the loop that `odo` uses for file synchronization:
+
+1. Deployment and Service resources are created (or updated) on the Kubernetes cluster
+2. Files are transfered over to the running container using the Kubernetes API 
+3. `odo` watches for file changes locally and repeats Step 1 and 2.
+
+The "loop" is cancelled once the user inputs Ctrl+C.
+
+#### Hot Reloading
+
+`hotReloadCapable` is a special boolean within `exec` that allows you to specify if a framework is "hot reloadable".
+
+If set to `true`, the container won't be restarted as the framework will handle file changes on its own.
+
+#### Full Example
+
+```yaml
+commands:
+  - exec:
+      commandLine: yarn dev
+      component: runtime
+      group:
+        isDefault: true
+        kind: run
+      hotReloadCapable: true
+      workingDir: ${PROJECT_SOURCE}
+    id: build
+```
+
+
+### What Commands are Executed in Dev or Deploy
+
+Each command has a group `kind`:
+
+```yaml
+- exec:
+    commandLine: npm start
+    component: runtime
+    group:
+      isDefault: true
+      kind: run
+    workingDir: ${PROJECT_SOURCE}
+  id: run
+```
+
+When `odo dev` is executed, we use: build, run, test and debug.
+
+These commands are typically ran within the `container` component that has been defined.
+
+TODO: When are the debug, test, run and build commands run? Should I just like.. #1 is build, then #2 is test (which isn't implemented), then #3 is run? #4 is debug but when you pass in --debug to `odo dev`? Should I do that?
+Ex.
+The order in which the commands are ran for `odo dev` are:
+1. `build`: We first build the necessary parts of the container for the application to correctly run
+2. `test`: NOT YET IMPLEMENTED?
+3. `run`: The application is ran within the container
+4. `debug`: This is ran when `odo dev --debug` is executed
+
+When `odo deploy` is executed, we use: deploy.
+
+These commands are typically tied to Kubernetes or OpenShift inline resources. They are defined as a component. However, you can use `container` or `image` components as well under the deploy group.
```

Comment:
You can also talk about building an image with an Image component, that will be used in some Kubernetes resource defined in Kubernetes component